libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1505,6 +1462,118 @@ internal class MapRouteLine(
                 else -> y2
             }
         }
+
+        /**
+         * This will extract all of the road classes for the various sections of the route
+         * and determine the distance from the origin point to the starting point for each
+         * road class section and associate that data with a traffic congestion value.
+         *
+         * Each item returned represents a distance from the origin point, the road class
+         * beginning at that distance and the traffic congestion value for that section. The
+         * road class and congestion remain until the next item in the collections.
+         *
+         * The items returned are ordered from the point closest to the origin to the point
+         * farthest from the origin. In other words from the beginning of the route until the end.
+         *
+         * @param route the the calculations should be performed on.
+         */
+        fun getRouteLineTrafficExpressionData(route: DirectionsRoute):
+            List<RouteLineTrafficExpressionData> {
+                val distances = route.legs()?.mapNotNull {
+                    it.annotation()?.distance()
+                }?.flatten()
+                ifNonNull(distances) { distanceList ->
+                    val roadClassMap = mutableMapOf<Int, String>()
+                    route.legs()?.asSequence()
+                        ?.mapNotNull { it.steps() }
+                        ?.flatten()
+                        ?.mapNotNull { it.intersections() }
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null && it.mapboxStreetsV8()?.roadClass() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] =
+                                it.mapboxStreetsV8()!!.roadClass()!!
+                        }
+
+                    var runningDistance = 0.0
+                    val routeLineTrafficData = mutableListOf<RouteLineTrafficExpressionData>()
+                    route.legs()?.mapNotNull {
+                        it.annotation()?.congestion()
+                    }?.flatten()?.forEachIndexed { index, congestion ->
+                        if (index == 0) {
+                            routeLineTrafficData.add(
+                                RouteLineTrafficExpressionData(
+                                    runningDistance,
+                                    congestion,
+                                    roadClassMap[index]
+                                )
+                            )
+                        } else {
+                            runningDistance += distanceList[index - 1]
+                            if (routeLineTrafficData.last().trafficCongestionIdentifier ==
+                                congestion &&
```

Comment:
Can we apply the same strategy of simplifying the expression also when road classes are not available, or not used for backfilling?

Reply:
It already works for both conditions. There are unit tests for the method using a routes with and without road classes. The test without road classes goes from 54 to 5 items in the result with that snippet of code.

Reply:
Cool! What I'm also looking for is removing the color duplicates from the expression itself, comment down below.

Also, which tests are impacted? I don't notice them in the diff.